### PR TITLE
Fix panicking when a new strand is created inside `lock`

### LINKED ISSUE #22987: Enable the following disabled tests
**Description:**
The following lock tests were disabled due to being intermittent. Please fix them and enable.

LocksInMainTest > simpleLock 
LocksInMainTest > testThrowErrorInsideLock
LocksInMainTest > testReturnInsideLock
LocksInMainTest > testThrowErrorInsideTryFinallyInsideLock
LocksInMainTest > testParallelRunUsingLocks
**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### LINKED ISSUE #28930: Intermittent test failure - ballerina daily build
**Description:**
Below test case is failing intermittently in daily build. Not able to reproduce it locally. Disabled the issue for now and required further analysis to drill down the root cause. Respective test case started failing from 25 days ago.

```
ballerina-new-parser-test-suite > jballerina-test > org.ballerinalang.test.lock.LocksInMainTest > testLocksWhenGlobalVariablesReferToSameValue FAILED
    org.ballerinalang.core.util.exceptions.BLangRuntimeException: Error while invoking function 'testLocksWhenGlobalVariablesReferToSameValue'
        at org.ballerinalang.test.BRunUtil.invoke(BRunUtil.java:447)
        at org.ballerinalang.test.BRunUtil.invokeFunction(BRunUtil.java:1253)
        at org.ballerinalang.test.BRunUtil.invoke(BRunUtil.java:195)
        at org.ballerinalang.test.BRunUtil.invoke(BRunUtil.java:1296)
        at org.ballerinalang.test.lock.LocksInMainTest.testLocksWhenGlobalVariablesReferToSameValue(LocksInMainTest.java:289)

        Caused by:
        java.lang.RuntimeException: Error while invoking function 'testLocksWhenGlobalVariablesReferToSameValue'

            Caused by:
            java.lang.reflect.InvocationTargetException

                Caused by:
                java.util.NoSuchElementException
```

### LINKED ISSUE #38047: [Bug]: `java.util.NoSuchElementException` at runtime in `BLockStore.panicIfInLock`
### Description

$title

```
Running executable

[2022-10-04 00:54:02,698] SEVERE {b7a.log.crash} - null 
java.util.NoSuchElementException
        at java.base/java.util.ArrayDeque.getLast(ArrayDeque.java:414)
        at io.ballerina.runtime.internal.BLock.lockedBySameContext(BLock.java:69)
        at io.ballerina.runtime.internal.BLockStore.panicIfInLock(BLockStore.java:58)
        at no_race_condition.main(no_race_condition.bal:10)
        at $_init.$lambda$main$(.)
        at io.ballerina.runtime.internal.scheduling.SchedulerItem.execute(SchedulerItem.java:54)
        at io.ballerina.runtime.internal.scheduling.Scheduler.run(Scheduler.java:324)
        at io.ballerina.runtime.internal.scheduling.Scheduler.runSafely(Scheduler.java:291)
        at java.base/java.lang.Thread.run(Thread.java:829)
```

### Steps to Reproduce

Maybe intermittent. Run several times to reproduce. Happens when the strands are on separate threads. Could not observe when they are on the same thread.
```bal
const rounds = 100;
const workers = 2;

int counter = 1;

public function main() {
    future<()> counterFuture = start check_counter();

    foreach int i in 0 ..< workers {
        _ = start increment_counter();
    }

    error? result = wait counterFuture;
}

function increment_counter() {
    foreach int i in 0 ..< rounds {
        lock {
            counter += 1;
        }
    }
}

function check_counter() {
    while (while_condition()) {}
}

function while_condition() returns boolean {
    lock {
        return counter < rounds * workers;
    }
}
```

### Affected Version(s)

Ballerina 2201.2.0 (Swan Lake Update 2)

### OS, DB, other environment details and versions

_No response_

### Related area

-> Runtime

### Related issue(s) (optional)

_No response_

### Suggested label(s) (optional)

_No response_

### Suggested assignee(s) (optional)

_No response_

### PR DESCRIPTION
**Pull request by Nadeeshan96**
$title
Fixes #38047 
Fixes #22987
Fixes #28930

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

**Review by HindujaB — APPROVED** [review]
LGTM

File: tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksInMainTest.java [hunk 1]
```diff
@@ -264,25 +134,29 @@ public void testLockNegativeCases() {
...
     @Test(description = "Test for parallel run when invocations have loops and global var dependencies")
     public void testParallelRunWithChainedInvocationDependencies() {
-        Object returns = BRunUtil.invoke(parallelCompileResult, "testLockWIthInvokableChainsAccessingGlobal");
+        BRunUtil.invoke(parallelCompileResult, "testLockWIthInvokableChainsAccessingGlobal");
     }
 
     @Test(description = "Test for parallel run when invocation are recursive and have global var dependencies")
     public void testParallelRunWithRecursiveInvocationDependencies() {
-        Object returns = BRunUtil.invoke(parallelCompileResult, "testLockWIthInvokableRecursiveAccessGlobal");
+        BRunUtil.invoke(parallelCompileResult, "testLockWIthInvokableRecursiveAccessGlobal");
```
Comment (gabilang) [inline tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksInMainTest.java:159]: Shall we add the tests in same `bal` files to `dataProvider`s?

Comment (Nadeeshan96) [reply tests/jballerina-unit-test/src/test/java/org/ballerinalang/test/lock/LocksInMainTest.java:159]: Each of these tests have a description associated with them. If I move them to use a `dataProvider`, those descriptions get omitted. Hence I did not do it like that.

**Review by warunalakshitha — APPROVED** [review]
LGTM

**Pull request merged**